### PR TITLE
DBZ-4375 Adding PubSub config options to pubsub connector

### PR DESCRIPTION
**Pull request by nathan-smit-1**
https://issues.redhat.com/browse/DBZ-4375

Currently, the Google Cloud Pub/Sub connector included with Debezium Server does not make use of any of the additional configuration options available to the Pub/Sub publisher.  Without additional options enabled, we have only been able to stream at ~300 rows per minute

Specifically, this change adds:

1) Batch settings:
https://cloud.google.com/pubsub/docs/samples/pubsub-publisher-batch-settings
2) Retry settings:
https://cloud.google.com/pubsub/docs/samples/pubsub-publisher-retry-settings
3) Flow Control:
https://cloud.google.com/pubsub/docs/samples/pubsub-publisher-flow-control

Ability to change these settings will greatly assist us in being able to tune the performance when pushing messages to the sink.

There are many configurable items for pubsub, so I've tried to focus on items that seem most important.

All the defaults and code are based on Google's Kafka --> Pubsub connector with the idea being that these are reasonable defaults that can then be configured further (  https://github.com/GoogleCloudPlatform/pubsub/tree/master/kafka-connector)

### CODE REVIEW

File: debezium-server/debezium-server-pubsub/src/main/java/io/debezium/server/pubsub/PubSubChangeConsumer.java [hunk 1]
```diff
@@ -54,6 +64,37 @@
 
     private static final String PROP_PREFIX = "debezium.sink.pubsub.";
     private static final String PROP_PROJECT_ID = PROP_PREFIX + "project.id";
+    private static final String PROP_BATCH_DELAY_THRESHOLD = PROP_PREFIX + "batch.settings.setDelayThreshold";
```
Comment (gunnarmorling) [inline debezium-server/debezium-server-pubsub/src/main/java/io/debezium/server/pubsub/PubSubChangeConsumer.java:67]: I don't think we need the "set" prefix"; and "settings" seems redundant, too? Also, the names should be adjusted to Debezium conventions, so this one should be `debezium.sink.pubsub.batch.delay.threshold.ms` (the `ms` suffix is common to all timeouts etc. given in milli-seconds).

File: debezium-server/debezium-server-pubsub/src/main/java/io/debezium/server/pubsub/PubSubChangeConsumer.java [hunk 2]
```diff
@@ -54,6 +64,37 @@
...
+    private Long maxOutstandingMessages;
+    private Long maxOutstandingRequestBytes;
+
+    private Integer maxTotalTimeoutMs;
+    private Integer maxRequestTimeoutMs;
+    private Integer initialRetryDelay;
+    private double retryDelayMultiplier;
+    private Long maxRetryDelay;
+    private Integer initialRpcTimeout;
+    private double rpcTimeoutMultiplier;
```
Comment (gunnarmorling) [inline debezium-server/debezium-server-pubsub/src/main/java/io/debezium/server/pubsub/PubSubChangeConsumer.java:97]: Could we obtain them via injection instead (as e.g. the case for `ordering.enabled`)? Should be a bit more concise?

File: debezium-server/debezium-server-pubsub/src/main/java/io/debezium/server/pubsub/PubSubChangeConsumer.java [hunk 3]
```diff
@@ -78,17 +119,59 @@
...
         if (customPublisherBuilder.isResolvable()) {
             publisherBuilder = customPublisherBuilder.get();
             LOGGER.info("Obtained custom configured PublisherBuilder '{}'", customPublisherBuilder);
             return;
         }
 
+        BatchingSettings.Builder batchingSettings = BatchingSettings.newBuilder()
+        .setDelayThreshold(Duration.ofMillis(maxDelayThresholdMs))
+        .setElementCountThreshold(maxBufferSize)
+        .setRequestByteThreshold(maxBufferBytes);
```
Comment (gunnarmorling) [inline debezium-server/debezium-server-pubsub/src/main/java/io/debezium/server/pubsub/PubSubChangeConsumer.java:150]: Can you please run the formatter via `mvn process-sources`.

**Review by gunnarmorling — COMMENTED** [review]
Thanks, that's a good addition. Some comments inline. In addition the options should also be added to the [documentation|https://debezium.io/documentation/reference/stable/operations/debezium-server.html#_google_cloud_pubsub] as part of this PR.

**Comment by gunnarmorling** [comment]
@nathan-smit-1, any news here?

**Comment by nathan-smit-1** [comment]
> @nathan-smit-1, any news here?

Hey @gunnarmorling, sorry for late response.  I've been working at this, just can't figure out how the injection works and ended up getting stuck.  Think it's just me not getting it.  Thought I could just add something like the below but in below case maxDelayThresholdMs value is just null

    @ConfigProperty(name = PROP_PREFIX + "batch.delay.threshold.ms", defaultValue = "100")
    Integer maxDelayThresholdMs;

Once I've figured out should be able to commit

File: debezium-server/debezium-server-pubsub/pom.xml [hunk 4]
```diff
@@ -30,7 +30,17 @@
...
+        <dependency>
+            <groupId>com.google.api</groupId>
+            <artifactId>gax</artifactId>
+            <version>1.56.0</version>
+        </dependency>
+        <dependency>
+            <groupId>org.threeten</groupId>
+            <artifactId>threetenbp</artifactId>
+            <version>1.4.3</version>
+        </dependency>
```
[the remark this answers is not in the document]
Comment (nathan-smit-1) [reply debezium-server/debezium-server-pubsub/pom.xml:42]: https://github.com/ThreeTen/threetenbp/blob/master/LICENSE.txt
https://github.com/googleapis/gax-java/blob/main/LICENSE

File: debezium-server/debezium-server-pubsub/src/main/java/io/debezium/server/pubsub/PubSubChangeConsumer.java (hunk 1 above)
Comment (nathan-smit-1) [reply debezium-server/debezium-server-pubsub/src/main/java/io/debezium/server/pubsub/PubSubChangeConsumer.java]: Adjusted in latest commit

File: debezium-server/debezium-server-pubsub/src/main/java/io/debezium/server/pubsub/PubSubChangeConsumer.java (hunk 2 above)
Comment (nathan-smit-1) [reply debezium-server/debezium-server-pubsub/src/main/java/io/debezium/server/pubsub/PubSubChangeConsumer.java]: Done.  I wasn't able to have the default be Long.MAX_VALUE using injection though so just used the actual Long.MAX_VALUE in the default if that makes sense.  Seems to work.

File: debezium-server/debezium-server-pubsub/src/main/java/io/debezium/server/pubsub/PubSubChangeConsumer.java (hunk 3 above)
Comment (nathan-smit-1) [reply debezium-server/debezium-server-pubsub/src/main/java/io/debezium/server/pubsub/PubSubChangeConsumer.java]: Done

**Review by nathan-smit-1 — COMMENTED** [review]
Have made adjustments based on feedback and commented in-line

**Comment by nathan-smit-1** [comment]
@gunnarmorling I've done a new commit with adjustments based on your feedback.  Also updated the debezium server documentation.  I gave the licenses inline but will paste here also:

https://github.com/ThreeTen/threetenbp/blob/master/LICENSE.txt
https://github.com/googleapis/gax-java/blob/main/LICENSE

File: debezium-server/debezium-server-pubsub/pom.xml (hunk 4 above)
[the remark this answers is not in the document]
Comment (gunnarmorling) [reply debezium-server/debezium-server-pubsub/pom.xml:42]: Ok, so both are BSD 3 clause, that's fine. Thx!

File: debezium-server/debezium-server-pubsub/src/main/java/io/debezium/server/pubsub/PubSubChangeConsumer.java [hunk 5]
```diff
@@ -85,10 +129,34 @@ void connect() {
             return;
         }
 
+        BatchingSettings.Builder batchingSettings = BatchingSettings.newBuilder()
+                .setDelayThreshold(Duration.ofMillis(maxDelayThresholdMs))
```
Comment (gunnarmorling) [inline debezium-server/debezium-server-pubsub/src/main/java/io/debezium/server/pubsub/PubSubChangeConsumer.java:133]: Wow, that's really unfortunate that they expose the 310 stuff in this signature :(

File: debezium-server/debezium-server-pubsub/pom.xml [hunk 6]
```diff
@@ -30,7 +30,17 @@
             <groupId>com.google.cloud</groupId>
             <artifactId>google-cloud-core</artifactId>
         </dependency>
-
+        <dependency>
+            <groupId>com.google.api</groupId>
+            <artifactId>gax</artifactId>
+            <version>1.56.0</version>
```
Comment (gunnarmorling) [inline debezium-server/debezium-server-pubsub/pom.xml:36]: Any reason for this seemingly old version? There's 2.8.1 as the latest?

**Review by gunnarmorling — COMMENTED** [review]
It's a pity that there is this 310 dependency, would have been so easy to avoid :/ But ok, I reckon things are as they are.

File: debezium-server/debezium-server-pubsub/pom.xml (hunk 6 above)
Comment (gunnarmorling) [reply debezium-server/debezium-server-pubsub/pom.xml:36]: @nathan-smit-1k, any thoughts on that one?

Btw. we generally prefer rebases over merges (I'll take care of that upon integrating that PR).

Comment (nathan-smit-1) [reply debezium-server/debezium-server-pubsub/pom.xml]: Hey @gunnarmorling .  Sorry about that, I did a fetch upstream and don't have separate branches (my fork is a bit disorganised).

I made a mistake on this.  I don't need to explicitly add these dependencies to the POM.xml file and it's probably a bad idea to have them there anyway so I've reverted the POM to its original state.  Both threetenbp and gax are actually dependencies of google-cloud-pubsub in maven so the version is related to the google-cloud-pubsub version which I believe is this one:  https://mvnrepository.com/artifact/com.google.cloud/google-cloud-pubsub/1.104.1 .  I've tested and everything compiles and runs fine

Comment (gunnarmorling) [reply debezium-server/debezium-server-pubsub/pom.xml]: Ok, makes sense. This LGTM now, will merge.

**Pull request merged**